### PR TITLE
sys/shell: refactor readline function

### PR DESCRIPTION
**Pull request by HendrikVE**
### Contribution description

With this PR I simply copied PR #12099 authored by @jcarrano. Since he left the team, I hereby offer to take care of rebasing and implement required changes.

This makes the code of readline() clearer and shorter. It also fixes a minor artifact of the long line handling.

Previously, it was not possible to recover from a long line. That is, if too many characters were sent, the line would be invalidated and pressing backspace would not fix it. The only option was to discard the line. It is now possible to bring the line back to size. Note that visual effects when deleting characters will still depend on the host's terminal.

The new code is written in a way that all writes to memory are guarded by bounds check, so an assertion was removed.

### Testing procedure

Test vectors were added to `tests/shell`. The test is not really testing a lot right now because the host is still line buffering.

### Issues/PRs references
Copy of #12099

#### Dependencies:
- [x] #13195

### CODE REVIEW

**Comment by jcarrano** [comment]
@HendrikVE thank you very much for taking over

**Comment by fjmolinas** [comment]
This one needs a rebase now!

**Comment by HendrikVE** [comment]
I'm already on it ;) The same with #13197

**Comment by HendrikVE** [comment]
Rebased and ready for review :)

**Comment by kaspar030** [comment]
I tried this on particle-xenon and nucleo-f072rb. Just typing a long line (hammering the keyboard) causes a failed assertio. Please take a look.

```
Welcome to RIOT!
> 
> 
> 
> 
> kjdföjasdfkljsdöljfölsdjföljsdöfjksdöaklfjlökasdjfölkasjdföljasdölfjasdlöjfölsdajfölasdjflöasdjölfjasdlökfjölas0x800110b
*** RIOT kernel panic:
FAILED ASSERTION.

        pid | name                 | state    Q | pri | stack  ( used) | base addr  | current     
          - | isr_stack            | -        - |   - |    512 (  132) | 0x20000000 | 0x200001c0
          1 | idle                 | pending  Q |  15 |    256 (  156) | 0x200003c0 | 0x20000424 
          2 | main                 | running  Q |   7 |   1536 (  672) | 0x200004c0 | 0x200008f4 
            | SUM                  |            |     |   2304 (  960)

*** halted.
```

**Comment by HendrikVE** [comment]
Thank you very much @kaspar030, my fixup commit hopefully fixed this issue.

**Comment by kaspar030** [comment]
Looking good now. I measure only 16b code increase, which is totally acceptable.

Please squash!

**Comment by kaspar030** [comment]
> Looking good now.

Sorry, backspace doesn't work anymore for me (partictle-xenon using picocom) :/

**Comment by HendrikVE** [comment]
That is a bug I apparently copied from #12099. Instead of looking for `0x7f (DEL)` `DEL` was defined to `0x07 (BEL)`. I fixed this and also rebased to recent master. Sorry for that.

**Comment by HendrikVE** [comment]
Well, tests are failing only for esp32, so it's not only my setup that is failing. Every time I enter data in to the shell the esp32 reboots with the following message

```
> help
help
ets Jun  8 2016 00:22:57

rst:0x7 (TG0WDT_SYS_RESET),boot:0x1b (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DOUT, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:4004
load:0x40078000,len:7304
load:0x40080000,len:7224
entry 0x40080350
W (29) boot: PRO CPU has been reset by WDT.
W (29) boot: WDT reset info: PRO CPU PC=0x400803c0
W (29) boot: WDT reset info: APP CPU PC=0x2fed1ff5

main(): This is RIOT! (Version: 2020.07-devel-17-g013af-shell-readline-refactor-new)
test_shell.
> 
```

Fun fact: As soon as I add a `printf("AAAAA");` in the first do-while loop of `handle_input_line` everything works fine, while a `printf("A");` is not sufficient. Really weird behaviour.

**Comment by HendrikVE** [comment]
It looks like a timing issue to me. I replaced my previous `printf` statement with a call of `xtimer_usleep` and it works for sleeping ~5ms. I checked this also for master and there I don't need to call xtimer. How is this possible? My PR did not even touch `handle_input_line` at all?

**Comment by fjmolinas** [comment]
> It looks like a timing issue to me. I replaced my previous printf statement with a call of xtimer_usleep and it works for sleeping ~5ms. I checked this also for master and there I don't need to call xtimer. How is this possible? My PR did not even touch handle_input_line at all?

@HendrikVE does the failure happen on master?

**Comment by HendrikVE** [comment]
@fjmolinas No, it does not. That is the point I don't understand. It works for master, but not for this PR based on the same master and it fails **after** my changes in `readline` at `handle_input_line`. When adding sleep of ~5ms it works as in master. I don't have any clue, especially because it seems to work fine for all other platforms...

**Comment by HendrikVE** [comment]
And adding `xtimer_usleep(5000);` to line 139 fixes the issue for some reason.

**Comment by fjmolinas** [comment]
> @fjmolinas No, it does not. That is the point I don't understand. It works for master, but not for this PR based on the same master and it fails after my changes in readline at handle_input_line. When adding sleep of ~5ms it works as in master. I don't have any clue, especially because it seems to work fine for all other platforms...

I don't have an `esp32` to attempt to reproduce, but I'll run the test on multiple boards, and see If I can get a board with the similar behavior.

**Comment by fjmolinas** [comment]
Maybe @gschorcht has some insight on why this might be failing on esp32?

**Comment by gschorcht** [comment]
> Maybe @gschorcht has some insight on why this might be failing on esp32?

Hm, I have absolutely no idea. I've been trying to debug the problem. It seems that the system crashes at this instruction: https://github.com/RIOT-OS/RIOT/blob/013af64a8034373dc92f95c2473a634f2719ee9f/sys/shell/shell.c#L197 But I'm not really sure, because the debugger also crashes at this point :worried: Even more strange, it works with with `-Og` or `-Os` but not with `-O2`.

**Comment by gschorcht** [comment]
I also took a look into the assembler code as generated by the compiler. Inspite of the fact that the code seems to be correct and should work, it leads to an exception at this instruction: https://github.com/RIOT-OS/RIOT/blob/013af64a8034373dc92f95c2473a634f2719ee9f/sys/shell/shell.c#L197

> Even more strange, it works with with `-Og` or `-Os` but not with `-O2`.

The differences between `-O2` and `-Os` are:
| Switch | `-O2` | `-Os` |
|:-------- | ------- | ------- |
| `-finline-functions` | disabled | enabled |
| `-foptimize-strlen` | enabled | disabled |
| `-fschedule-insns` | enabled | disabled |

So I tried different compiler switches to reproduce the problem. Unfortunatly, using `CFLAGS_OPT='-O2 -finline-functions'` doesn't work while `CFLAGS_OPT='-Os -foptimize-strlen -fschedule-insns'` still works.

We changed the optimization flags from `-Os` to `-O2` because of a sporadic problem we had with `tests/bench*`: https://github.com/RIOT-OS/RIOT/blob/3e922f878a1f713530efe38ec154de916002ee34/cpu/esp32/Makefile.include#L10-L12

We could try changing it back to `-Os` and see if `tests/bench*` are still a problem.

**Comment by kaspar030** [comment]
> it leads to an exception at this instruction:
> 
> https://github.com/RIOT-OS/RIOT/blob/013af64a8034373dc92f95c2473a634f2719ee9f/sys/shell/shell.c#L197

Hm, that's a write access to a newly created VLA. The index is certainly correct (there must be an `argv[argc]` if the array size is `argc + 1`). Are we 100% sure argc is correct? Is there enough stack space?

**Comment by kaspar030** [comment]
`CTRL-D` on native seems to loop printing the prompt?

**Comment by fjmolinas** [comment]
> CTRL-D on native seems to loop printing the prompt?

Same as  https://github.com/RIOT-OS/RIOT/pull/13675 right?

**Comment by gschorcht** [comment]
> > it leads to an exception at this instruction:
> > https://github.com/RIOT-OS/RIOT/blob/013af64a8034373dc92f95c2473a634f2719ee9f/sys/shell/shell.c#L197
> 
> Hm, that's a write access to a newly created VLA. The index is certainly correct (there must be an `argv[argc]` if the array size is `argc + 1`).

> Are we 100% sure argc is correct?

It seems so, in debugger I saw that `argc` was 1.

> Is there enough stack space?

For the ESP32 there was also enough stack space and 16 bytes were allocated for `argv[argc + 1]` in that case which was more than enough. Although only 8 bytes were required, 16 bytes were probably allocated due to alignment requirements for performance reasons.

**Comment by fjmolinas** [comment]
> We could try changing it back to -Os and see if tests/bench* are still a problem.

I don't have experience with `esp32` to know whats best here, and I have no `esp32` to dig in further :/

**Comment by HendrikVE** [comment]
I have an esp32 here, but unfortunately I don't have enough knowledge about the toolchain to say what the consequences and possible side effects are when switching optimization flags. @gschorcht , is `tests/bench*` still a problem when using `-Os`?`If so, does the esp32 hang completely or just for a few ms or something?

**Comment by gschorcht** [comment]
> @gschorcht , is `tests/bench*` still a problem when using `-Os`?` If so, does the esp32 hang completely or just for a few ms or something?

I don't know, I have to try again.

**Comment by gschorcht** [comment]
> > @gschorcht , is `tests/bench*` still a problem when using `-Os`?` If so, does the esp32 hang completely or just for a few ms or something?
> 
> I don't know, I have to try again.

I could not reproduce the problem again, at least not at the moment. It was just a problem that occurred sporadically. However, when it did occur, the ESP32 hang up completely.

At that time I investigated the problem and found that there was a hardware sync problem when reading the hardware timer in `benchmark_print_time` if the interrupts were disabled during the benchmark test.

BTW, I never saw this problem in CI, but @kaspar30 modified the RESET logic of the ESP32 board a bit for the CI.

**Comment by kaspar030** [comment]
> BTW, I never saw this problem in CI, but @kaspar30 modified the RESET logic of the ESP32 board a bit for the CI.

I added a cap that was missing on the cheap china clones CI is using.

**Comment by fjmolinas** [comment]
> I could not reproduce the problem again, at least not at the moment. It was just a problem that occurred sporadically. However, when it did occur, the ESP32 hang up completely.

@gschorcht  What if we just change the optimization parameter for that test then? `CFLAGS_OPT` can be overriden, and that way we unblock this PR :)

**Comment by gschorcht** [comment]
> @gschorcht What if we just change the optimization parameter for that test then? `CFLAGS_OPT` can be overriden, and that way we unblock this PR :)

Yeah, that would be one possibility. But as I said, the problem occurred sporadically only and I couldn't reproduce it again. So I'm thinking about to change the optimization option back to `-Os` and to observe the test runs in CI to see whether it occurs again.

Maybe, we could introduce this change piggybacked with this PR.

**Comment by fjmolinas** [comment]
> Maybe, we could introduce this change piggybacked with this PR.

@HendrikVE can you add a commit changing the optimization back to `Os`?

**Comment by fjmolinas** [comment]
> @HendrikVE can you add a commit changing the optimization back to Os?

With a commit message reflecting what @gschorcht has mentioned :)

**Comment by gschorcht** [comment]
> > Maybe, we could introduce this change piggybacked with this PR.
> 
> @HendrikVE can you add a commit changing the optimization back to `Os`?

I can inform you of this change later today. I am sorry for the delayed reaction at the moment, but I am completely busy preparing my online courses for the virtual online summer term at the university.

**Comment by fjmolinas** [comment]
> I can provide this change later today. I am sorry for the delayed reaction at the moment, but I am completely busy preparing my online courses for the virtual online summer term at the university.

I think its not bad to do it in this PR, to test that the change actually fixes the issue here.

**Comment by fjmolinas** [comment]
> I think its not bad to do it in this PR, to test that the change actually fixes the issue here.

I just pushed a commit doing it (hope that is fine @HendrikVE), lets see if the CI is happy with that. @gschorcht let us know if the commit message in 333fb75 is good for you, otherwise I can amend (or @HendrikVE )

**Comment by fjmolinas** [comment]
Everything is green now @kaspar030 you had been reviewing this PR so far, are you fine with the changes?

@gschorcht are you with the last commit?

File: cpu/esp32/Makefile.include [hunk 1]
```diff
@@ -7,9 +7,7 @@ endif
 
 ESP_SDK_DIR = $(ESP32_SDK_DIR)
 
-# With the '-Os' option, the ESP32 hangs sporadically in 'tests/bench*' if
-# interrupts are disabled too early by benchmark tests.
-CFLAGS_OPT ?= -O2
+CFLAGS_OPT ?= -Os
```
Comment (gschorcht) [inline cpu/esp32/Makefile.include:10]:
```suggestion
```
`CFLAGS_OPT` is set to `-Os` by default in https://github.com/RIOT-OS/RIOT/blob/8ffd34ddd05627e44193945e8365aeca07899875/cpu/esp_common/Makefile.include#L45

File: cpu/esp32/Makefile.include [hunk 2]
```diff
@@ -7,9 +7,7 @@ endif
 
 ESP_SDK_DIR = $(ESP32_SDK_DIR)
```
Comment (gschorcht) [inline cpu/esp32/Makefile.include:9]:
```suggestion

```
And remove the additional blank line. Unfortunately, multiline suggestions are not supported for deleted lines :worried:

**Comment by HendrikVE** [comment]
Thanks @fjmolinas and @gschorcht. I amended the commit and its message about the optimization flag.

**Comment by kaspar030** [comment]
Gave this a quick test, looking fine now.

**Comment by fjmolinas** [comment]
@HendrikVE can you test if the failing test fails in master as well? I saw it fail in another PR

**Comment by gschorcht** [comment]
> @HendrikVE can you test if the failing test fails in master as well? I saw it fail in another PR

This problem definitely has nothing to do with this PR, it only occurs sporadically in Murdock. `tests/pkg_littlefs` work perfect on my local system. Maybe it is a hardware problem with the flash memory on certain workers. I'm trying to figure out whether the problem occurs on certain workers. Til now, I have observed the problem on `pi-6356c652` and `pi-67e8efb7`.

Let's rerun the test and see what happens.

**Comment by fjmolinas** [comment]
There is something wrong with `esp32-wroom32` in master, its failing multiple tests, and has been failing multiple tests in nightly for the last couple of days.

**Comment by kaspar030** [comment]
> There is something wrong with `esp32-wroom32` in master, its failing multiple tests, and has been failing multiple tests in nightly for the last couple of days.

It seems like not only esp32-wroom32 is failing tests though. :(

**Comment by kaspar030** [comment]
> It seems like not only esp32-wroom32 is failing tests though. :(

sorry, I think I misread the output, you're right.

**Comment by gschorcht** [comment]
Yeah, I already saw it in the nightlies. According to the nightlies the tests fail since May 10th. It seems to be related to the fix in PR #14041 that was merged on May, 9th. This fix sets `BUILD_BEFORE_FLASH` to ensure that the ELF file is build before the flash image is generated. Is this variable used by the CI?

**Comment by kaspar030** [comment]
> Is this variable used by the CI?

Probably now it is used too much ;) CI splits building and testing. One (fast) worker compiles the elf and sends it to a queue as "test job", then a RasPi picks that, copies the .elf to its place and calls "make test". it seems that due to that change, it tries to rebuild the .elf, or something in that direction.

**Comment by kaspar030** [comment]
Let's continue in #14041!

**Comment by fjmolinas** [comment]
Hopefully this build will the the last..

**Comment by HendrikVE** [comment]
Looks good now!

**Comment by HendrikVE** [comment]
Cool! Thanks a lot @kaspar030, @fjmolinas and @gschorcht for your reviews! #13197 needs a rebase now, I'll do that soon.

**Pull request merged**